### PR TITLE
[Publishing] Add Publish to S3 using Buildkite

### PR DESCRIPTION
**Pull request by ParaskP7**
App Infra Request: pdnsEh-1RV-p2

Depends On:
- [x] Security related inquiry/discussion on p1727693975599849-slack-C02KLTL3MKM.
- [x] [BuildkiteCI#526](https://github.com/Automattic/buildkite-ci/pull/526)
- [x] Configure `.mobile-secrets` (create deployment key & environmental variables),  and `GitHub` (add deployment key).

### Description

This PR adds publishing to our S3 Maven using Buildkite:

### Testing Instructions

`TODO`:
- Wait for this [BuildkiteCI#526](https://github.com/Automattic/buildkite-ci/pull/526) PR to get deployed/merged.
- Trigger a CI build on this PR and wait for the library to be published on our S3 Maven.
- Use a project like [Gravatar-SDK-Android](https://github.com/Automattic/Gravatar-SDK-Android) to update the group from `com.github.yalantis` to `com.automattic`, while using a `<PR>-<COMMIT_SHA>` combination as its test version.
- Build the test project and verify that any `uCrop` related functionality is still working as expected.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -195,6 +186,14 @@ We’d be really happy if you sent us links to your projects where you use our c
 - [BCReader](https://play.google.com/store/apps/details?id=com.iac.bcreader).
 - [Xprezia: Share Your Passion](https://play.google.com/store/apps/details?id=com.xprezzia.cnj).
 
+## Publishing a new version
+
+In the following cases, the CI will publish a new version with the following format to our remote Maven repo:
+
+* For each commit in an open PR: `<PR-number>-<commit full SHA1>`
+* Each time a PR is merged to `trunk`: `trunk-<commit full SHA1>`
```
Comment (AdamGrzybkowski) [inline README.md]: Are we renaming `develop` to `trunk`? I probably forgot to add one piece of information in the request - this repo has two important branches `develop` and `develop-non-native`. The one that we're interested in and planning to use is the second non-native one.

Comment (ParaskP7) [reply README.md]: Ah apologies @AdamGrzybkowski , this is a copy-pasting mistake, which I took from similar instructions we have on every other library of ours, I'll go ahead and rename `trunk` to `develop-non-native` on this `README.md` file.

> I probably forgot to add one piece of information in the request - this repo has two important branches develop and develop-non-native. The one that we're interested in and planning to use is the second non-native one.

Yea, I noticed that when I cloned the repo, seeing `develop-non-native` being the default branch, and so, I based everything on that, while ignoring `develop` for now. Thanks for the clarification! 🙇

Comment (ParaskP7) [reply README.md]: Also, now that we're talking about this, maybe I should delete this line as well, wdyt?

```
``` implementation 'com.automattic:ucrop:2.2.10-native' ``` - get power of the native code to preserve image quality (+ about 1.5 MB to an apk size)
```

And only keep this one line there:

```
``` implementation 'com.automattic:ucrop:2.2.10' ``` - lightweight general solution
```

Just to make it clear that we are only dealing/publishing the `non-native` side of things. 🤔

Comment (AdamGrzybkowski) [reply README.md]: Sounds good to me 👍

Comment (ParaskP7) [reply README.md]: This is now done: 404636a

**Review by wzieba — COMMENTED** [review]
Looks good to me! Let's see results when CI will be triggered

File: README.md [hunk 2]
```diff
@@ -15,18 +15,7 @@
...
-	   }
-	}
-	```
-
-    ``` implementation 'com.github.yalantis:ucrop:2.2.9' ``` - lightweight general solution
-
-    ``` implementation 'com.github.yalantis:ucrop:2.2.9-native' ``` - get power of the native code to preserve image quality (+ about 1.5 MB to an apk size)
+    ``` implementation 'com.automattic:ucrop:2.2.10' ``` - lightweight general solution
 
 2. Add UCropActivity into your AndroidManifest.xml
```
Comment (wzieba) [inline README.md:20]: It's not a scope of this PR, but I think that the SDK should add this activity to its own manifest and rely on manifest merge to include it into client's manifest.

**Comment by ParaskP7** [comment]
**UPDATE**:

1. After configuring `.mobile-secrets` and GitHub, I run this [first](https://buildkite.com/automattic/ucrop/builds/1) build, which verified that the CI setup has been successful (🎉) but nevertheless `Publish to S3` failed with an out of memory exception. ❌
2. After updating `org.gradle.jvmargs` (see f881f29) the [second](https://buildkite.com/automattic/ucrop/builds/2#0192486a-0c78-4b5a-abed-bd4b343308f8) manual build ended-up in a total success (🎉) and our own version of `uCrop` got published to `s3` for the first time: ✅ 

```
petros.paraskevopoulos@Mac-Studio uCrop % aws s3 ls a8c-libs/android/com/automattic/ucrop/
                           PRE publishing_add-publish-to-s3-using-buildkite-f881f29c6a1802963b98c68f44b0b840b2417e4d/
2024-10-01 17:17:04        566 maven-metadata.xml
2024-10-01 17:17:04         32 maven-metadata.xml.md5
2024-10-01 17:17:04         40 maven-metadata.xml.sha1
2024-10-01 17:17:04         64 maven-metadata.xml.sha256
2024-10-01 17:17:04        128 maven-metadata.xml.sha512
```

3. I'll now try and see if I can use this new `<PR>-<COMMIT_SHA>` combination on the [Gravatar-SDK-Android](https://github.com/Automattic/Gravatar-SDK-Android) repo and that it compiles/works as expected. 🤞

**Comment by ParaskP7** [comment]
**UPDATE** on [(3.)](https://github.com/Automattic/uCrop/pull/2#issuecomment-2386121149) above:

I tested our version of `uCrop`, the `Automattic/uCrop`, on [Gravatar-SDK-Android](https://github.com/Automattic/Gravatar-SDK-Android), using a `<BRANCH>-<COMMIT_SHA>` instead (the `<PR>-<COMMIT_SHA>` combination wasn't available, it probably being a manual, and not a PR build).

Result: Everything is working as expected! ✅ 

Below is a patch for you to try it yourself @AdamGrzybkowski, just to verify that everything is working as expected with this version of `uCrop`. Then, we could merge this and create a new release, as per the `README.md`, the `2.2.10`, or else, the `com.automattic:ucrop:2.2.10`, and start using it:

<details>
    <summary>Batch</summary>

```diff
Index: gradle/libs.versions.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/gradle/libs.versions.toml b/gradle/libs.versions.toml
--- a/gradle/libs.versions.toml	(revision 345e88842413555037409b599a5d9aaec2081ed4)
+++ b/gradle/libs.versions.toml	(date 1727793809388)
@@ -15,7 +15,7 @@
 mockk = "1.13.9"
 appcompat = "1.7.0"
 material = "1.12.0"
-ucrop = "2.2.9"
+ucrop = "publishing_add-publish-to-s3-using-buildkite-f881f29c6a1802963b98c68f44b0b840b2417e4d"
 coil = "2.7.0"
 activityCompose = "1.9.2"
 robolectric = "4.12.2"
@@ -64,7 +64,7 @@
 retrofit = { group = "com.squareup.retrofit2", name = "retrofit", version.ref = "retrofit" }
 retrofit-moshi-converter = { group = "com.squareup.retrofit2", name = "converter-moshi", version.ref = "retrofit" }
 okhttp = { group = "com.squareup.okhttp3", name = "okhttp", version.ref = "okhttp" }
-ucrop = { group = "com.github.yalantis", name = "ucrop", version.ref = "ucrop" }
+ucrop = { group = "com.automattic", name = "ucrop", version.ref = "ucrop" }
 coil-compose = { group = "io.coil-kt", name = "coil-compose", version.ref = "coil" }
 coil-svg = { group = "io.coil-kt", name = "coil-svg", version.ref = "coil" }
 coil-test = { group = "io.coil-kt", name = "coil-test", version.ref = "coil" }
Index: settings.gradle.kts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/settings.gradle.kts b/settings.gradle.kts
--- a/settings.gradle.kts	(revision 345e88842413555037409b599a5d9aaec2081ed4)
+++ b/settings.gradle.kts	(date 1727793854753)
@@ -26,6 +26,13 @@
         google()
         mavenCentral()
         maven { url = uri("https://jitpack.io") }
+        maven {
+            url = uri("https://a8c-libs.s3.amazonaws.com/android")
+            content {
+                includeGroup("com.automattic")
+                includeGroup("com.automattic.ucrop")
+            }
+        }
     }
 }
```

</details>

PS: Also, from my screenshot below you will notice that we are indeed bringing the correct version of `BitmapCropTask`, the one with the changes you need:

image

Cc @wzieba

**Comment by AdamGrzybkowski** [comment]
>Below is a patch for you to try it yourself @AdamGrzybkowski, just to verify that everything is working as expected with this version of uCrop.

Thanks for the patch - I tested it and all worked fine!

**Review by AdamGrzybkowski — APPROVED** [review]

**Comment by ParaskP7** [comment]
Awesome, thanks for testing @AdamGrzybkowski , I'll proceed with merging this soon and then will try publishing our new `om.automattic:ucrop:2.2.10` release, for it to then be uploaded to S3, hoping that all the tooling work I did so far would suffice. 🤞

**Pull request merged**